In [6]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [7]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [8]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [9]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [18]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
df_market_scaled = StandardScaler().fit_transform(df_market_data
                        [["price_change_percentage_24h","price_change_percentage_7d","price_change_percentage_14d",
                          "price_change_percentage_30d","price_change_percentage_60d",
                          "price_change_percentage_200d","price_change_percentage_1y"]])
df_market_scaled

array([[ 5.08529366e-01,  4.93193071e-01,  7.72200433e-01,
         2.35459633e-01, -6.74950963e-02, -3.55953481e-01,
        -2.51636882e-01],
       [ 1.85445894e-01,  9.34445040e-01,  5.58692121e-01,
        -5.43409317e-02, -2.73482725e-01, -1.15759474e-01,
        -1.99352110e-01],
       [ 2.17739616e-02, -7.06336853e-01, -2.16804207e-02,
        -6.10301536e-02,  8.00452481e-03, -5.50246924e-01,
        -2.82060506e-01],
       [-4.07643829e-02, -8.10928066e-01,  2.49457974e-01,
        -5.03879651e-02, -3.73164019e-01, -4.58258816e-01,
        -2.95546142e-01],
       [ 1.19303608e+00,  2.00095907e+00,  1.76061001e+00,
         5.45842065e-01, -2.91202870e-01, -4.99847761e-01,
        -2.70316950e-01],
       [ 8.91870708e-01,  1.32729453e+00,  8.00214184e-01,
        -5.71478992e-02,  7.78653106e-01, -1.88231917e-01,
        -2.25532605e-01],
       [ 1.13972400e-02,  2.57225091e+00,  1.10164693e+00,
        -4.90495415e-01, -9.31954023e-01,  3.87758986e-01,
        -1.8284399

In [28]:
# Create a DataFrame with the scaled data
df_market_data_scaled = pd.DataFrame (df_market_scaled,columns = ["price_change_percentage_24h","price_change_percentage_7d","price_change_percentage_14d",
                          "price_change_percentage_30d","price_change_percentage_60d",
                          "price_change_percentage_200d","price_change_percentage_1y"] )
df_market_data_scaled
# Copy the crypto names from the original data
coins = list(df_market_data.index) 
coins
# Set the coinid column as index
df_market_data_scaled['coin_id'] = coins #creating  a new column in the scaled df containing the list of coins
df_market_data_scaled = df_market_data_scaled.set_index('coin_id') #setting  the df index to the new coin_id column

# Display sample data
df_market_data_scaled 

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                         
bitcoin                                  0.508529                    0.493193   
ethereum                                 0.185446                    0.934445   
tether                                   0.021774                   -0.706337   
ripple                                  -0.040764                   -0.810928   
bitcoin-cash                             1.193036                    2.000959   
binancecoin                              0.891871                    1.327295   
chainlink                                0.011397                    2.572251   
cardano                                  0.102530                    1.508001   
litecoin                                 0.077497                    0.334297   
bitcoin-cash-sv                          0.448952                   -0.190684   
crypto-com-chain                         0.331280                   -1.614844   
usd-coin                                 0.034352                   -0.733026   
eos                                      0.155710                   -0.922491   
monero                                   0.262723                    1.792602   
tron                                     0.130050                   -0.041018   
tezos                                   -0.151583                    0.708196   
okb                                     -0.923203                   -1.437359   
stellar                                 -0.277543                   -0.385209   
cosmos                                  -0.255978                    1.840274   
cdai                                     0.180851                   -0.704931   
neo                                      0.286546                   -0.326301   
wrapped-bitcoin                          0.515453                    0.461843   
leo-token                                0.051758                   -0.928381   
huobi-token                             -0.052032                   -0.457229   
nem                                     -0.217984                   -0.849381   
binance-usd                              0.061339                   -0.706669   
iota                                     0.259097                    0.249508   
vechain                                  0.585089                   -0.994231   
zcash                                   -0.127467                    0.929119   
theta-token                             -1.612188                   -1.682027   
dash                                    -0.296940                    0.094763   
ethereum-classic                        -0.071312                   -0.229484   
ethlend                                 -4.981042                   -0.045178   
maker                                   -0.125168                    0.580730   
havven                                  -1.428574                   -0.025510   
omisego                                  1.919812                    0.370447   
celsius-degree-token                     1.045530                   -0.618328   
ontology                                -0.409044                   -0.906963   
ftx-token                                0.414711                    0.414044   
true-usd                                 0.078038                   -0.687745   
digibyte                                 1.217453                   -0.607714   

                      price_change_percentage_14d  \
coin_id                                             
bitcoin                                  0.772200   
ethereum                                 0.558692   
tether                                  -0.021680   
ripple                                   0.249458   
bitcoin-cash                             1.760610   
binancecoin                              0.800214   
chainlink                                1.101647   
cardano                                  0.648885   
litecoin

---

### Find the Best Value for k Using the Original Data.

In [30]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1,12))
k

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

In [32]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
for i in k:
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
    k_model = KMeans(n_clusters=i, random_state=1)
# 2. Fit the model to the data using `df_market_data_scaled`
    k_model.fit(df_market_data_scaled)
# 3. Append the model.inertia_ to the inertia list
    inertia.append(k_model.inertia_)

C:\Users\abc\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
C:\Users\abc\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
C:\Users\abc\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
C:\Users\abc\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid

In [33]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data ={"k":k, "inertia": inertia}

# Create a DataFrame with the data to plot the Elbow curve
elbow_df = pd.DataFrame(elbow_data)
elbow_df

k     inertia
0    1  287.000000
1    2  195.820218
2    3  123.190482
3    4   79.022435
4    5   65.405923
5    6   52.933559
6    7   47.983124
7    8   37.288187
8    9   33.061685
9   10   28.779752
10  11   25.248499

In [36]:
# Plot a line chart with all the inertia values computed with
elbow_df.hvplot.line(
    x="k",
    y="inertia",
    title ="Elbow Curve",
    xticks=k
)
# the different values of k to visually identify the optimal value for k.
# K:3 ,K:4,K:6

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:4 

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [37]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state =1)

In [38]:
# Fit the K-Means model using the scaled data
model.fit(df_market_data_scaled)

C:\Users\abc\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
C:\Users\abc\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4, random_state=1)

In [39]:
# Predict the clusters to group the cryptocurrencies using the scaled data
k_4 = model.predict(df_market_data_scaled)

# Print the resulting array of cluster values.
k_4

array([3, 3, 1, 1, 3, 3, 3, 3, 3, 1, 1, 1, 1, 3, 1, 3, 1, 1, 3, 1, 1, 3,
       1, 1, 1, 1, 1, 1, 3, 1, 1, 1, 0, 3, 1, 1, 2, 1, 1, 1, 1])

In [40]:
# Create a copy of the DataFrame
clusters_predicted = df_market_data_scaled.copy()

In [42]:
# Add a new column to the DataFrame with the predicted clusters
clusters_predicted['predicted_cluster']=k_4


# Display sample data
clusters_predicted.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  predicted_cluster  
coin_id                                                      
bitcoin                        -0.251637                  3  
ethereum                       -0.199352                  3  
tether                         -0.282061                  1  
ripple                         -0.295546                  1  
bitcoin-cash                   -0.270317                  3

In [43]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
clusters_predicted.hvplot.scatter(
    x='price_change_percentage_24h',
    y='price_change_percentage_7d',
    by = 'predicted_cluster',
    hover_cols = 'coin_id'
)

:NdOverlay   [predicted_cluster]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [45]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components = 3)

In [46]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
clusters_pca = pca.fit_transform(clusters_predicted)

# View the first five rows of the DataFrame. 
clusters_pca[:5]

array([[-1.58513942,  0.13439444,  0.545621  ],
       [-1.32258066,  0.05449393,  1.11570854],
       [ 0.26633536, -0.53640937, -0.79694588],
       [ 0.25722513, -0.59201286, -0.6465287 ],
       [-2.95224985,  0.42237155,  1.26789353]])

In [47]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
pca.explained_variance_ratio_

array([0.37005408, 0.32322221, 0.19115222])

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:0.37005408+0.32322221+0.19115222 = 0.8844285111826466 (that means with 3 principal components over 88percent of the variance in data is captured in PC1, PC2, PC3)

In [50]:
# Create a new DataFrame with the PCA data.
# Creating a DataFrame with the PCA data
df_market_data_pca = pd.DataFrame(
    clusters_pca,
    columns = ["PC1","PC2","PC3"]
)
df_market_data_pca




# Copy the crypto names from the original data

df_market_data_pca['coin_id'] = list(clusters_predicted.index)
df_market_data_pca
# Set the coinid column as index
df_market_data_pca = df_market_data_pca.set_index('coin_id')

# Display sample data
df_market_data_pca.head()

PC1       PC2       PC3
coin_id                                   
bitcoin      -1.585139  0.134394  0.545621
ethereum     -1.322581  0.054494  1.115709
tether        0.266335 -0.536409 -0.796946
ripple        0.257225 -0.592013 -0.646529
bitcoin-cash -2.952250  0.422372  1.267894

---

### Find the Best Value for k Using the PCA Data

In [52]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1,12))
k

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

In [53]:
# Create an empty list to store the inertia values
inertia2 =[]

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for i in k:
    k_model = KMeans(n_clusters=i, random_state= 1)
    k_model.fit(df_market_data_pca)
    inertia2.append(k_model.inertia_)

C:\Users\abc\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
C:\Users\abc\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
C:\Users\abc\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
C:\Users\abc\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid

In [54]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data2 = {"k":k, "inertia":inertia2}
# Create a DataFrame with the data to plot the Elbow curve
elbow_df2 = pd.DataFrame(elbow_data2)
elbow_df2

k     inertia
0    1  287.007838
1    2  196.268358
2    3  111.972027
3    4   43.205879
4    5   33.042435
5    6   24.347640
6    7   19.777320
7    8   15.868897
8    9   12.351058
9   10   10.169351
10  11    8.095958

In [55]:
# Plot a line chart with all the inertia values computed with 
elbow_df2.hvplot.line(
    x="k",
    y="inertia",
    title = "Elbow Curve",
)
# the different values of k to visually identify the optimal value for k.


:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** The best k-value is k=4 when using PCA data


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** No, it is the same k value as found using the original data

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [56]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=1)

In [57]:
# Fit the K-Means model using the PCA data
model.fit(df_market_data_pca)

C:\Users\abc\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
C:\Users\abc\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4, random_state=1)

In [58]:
# Predict the clusters to group the cryptocurrencies using the PCA data
k_4 = model.predict(df_market_data_pca)
# Print the resulting array of cluster values.
k_4

array([1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 1, 0, 0, 3, 0, 0, 0, 0])

In [59]:
# Create a copy of the DataFrame with the PCA data
copy_df_market_data_pca = df_market_data_pca.copy()
# Add a new column to the DataFrame with the predicted clusters
copy_df_market_data_pca['predicted_cluster'] = k_4

# Display sample data
copy_df_market_data_pca.head()

PC1       PC2       PC3  predicted_cluster
coin_id                                                      
bitcoin      -1.585139  0.134394  0.545621                  1
ethereum     -1.322581  0.054494  1.115709                  1
tether        0.266335 -0.536409 -0.796946                  0
ripple        0.257225 -0.592013 -0.646529                  0
bitcoin-cash -2.952250  0.422372  1.267894                  1

In [60]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
copy_df_market_data_pca.hvplot.scatter(
    x="PC1",
    y="PC2",
    by ='predicted_cluster',
    hover_cols ='coin_id'
)

:NdOverlay   [predicted_cluster]
   :Scatter   [PC1]   (PC2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [66]:
# Composite plot to contrast the Elbow curves
elbow_df.hvplot.line(x="k", y="inertia", title= "Elbow Curve 1", xticks=k) + elbow_df2.hvplot.line(x="k", y="inertia", title="Elbow Curve 2", xticks=k)

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [67]:
# Composite plot to contrast the clusters
clusters_predicted.hvplot.scatter( x='price_change_percentage_24h', y='price_change_percentage_7d',by='predicted_cluster', hover_cols = 'coin_id') + copy_df_market_data_pca.hvplot.scatter(x="PC1", y="PC2", by = 'predicted_cluster', hover_cols='coin_id')

:Layout
   .NdOverlay.I  :NdOverlay   [predicted_cluster]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [predicted_cluster]
      :Scatter   [PC1]   (PC2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** The impact of using PCA data resulted in low inertia and high concentration which means each cluster is small                   and well spaced from other clusters and also resulted in more entries within cluster 0 and  cluster 1 wheras if                 we see it on actual dataset there is high inertia and low concentration as compared .Hence using PCA results in                 better analysis.
  